metaserver/src/main/scala/scala/meta/languageserver/Compiler.scala

```diff
@@ -102,27 +105,38 @@ class Compiler(
     Effects.InstallPresentationCompiler
   }
 
-  private def indexDependencyClasspath(
-      config: CompilerConfig
+  // NOTE(olafur) this probably belongs somewhere else than Compiler.
+  def indexDependencyClasspath(
+      sourceJars: List[AbsolutePath]
   ): Effects.IndexSourcesClasspath = {
     if (!serverConfig.indexClasspath) return Effects.IndexSourcesClasspath
+    val sourceJarsWithJDK =
+      if (serverConfig.indexJDK)
+        CompilerConfig.jdkSources.fold(sourceJars)(_ :: sourceJars)
+      else sourceJars
     val buf = List.newBuilder[AbsolutePath]
-    val sourceJars = config.sourceJars
-    sourceJars.foreach { jar =>
+    sourceJarsWithJDK.foreach { jar =>
       // ensure we only index each jar once even under race conditions.
+      // race conditions are not unlikely since multiple .compilerconfig
+      // are typically created at the same time for each project/configuration
+      // combination. Duplicate tasks are expensive, for example we don't want
+      // to index the JDK twice on first startup.
       indexedJars.computeIfAbsent(jar, _ => buf += jar)
     }
-    val sourcesClasspath = buf.result()
-    if (sourcesClasspath.nonEmpty) {
-      logger.info(
-        s"Indexing classpath with ${sourcesClasspath.length} entries..."
-      )
-    }
-    ctags.Ctags.index(sourcesClasspath) { doc =>
-      documentSubscriber.onNext(doc)
+    val sourceJarsToIndex = buf.result()
+    // Acquire a lock on the leveldb cache only during indexing.
+    LevelDBMap.withDB(cacheDirectory.resolve("leveldb").toFile) { db =>
```

Comment:
Is there any value in taking smaller locks during each individual indexDatabase call to give others time to jump in the gaps if it takes a long time to load indexes?

Reply:
There is, I've opened https://github.com/scalameta/language-server/issues/49 to track improvements on this.